source/common/network/io_socket_handle_impl.cc

```diff
@@ -502,6 +512,46 @@ Api::IoCallUint64Result IoSocketHandleImpl::recvmmsg(RawSliceArrays& slices, uin
 }
 
 Api::IoCallUint64Result IoSocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if constexpr (Event::PlatformDefaultTriggerType == Event::FileTriggerType::EmulatedEdge) {
+    // Mimic the MSG_PEEK by read data.
+    if (flags & MSG_PEEK) {
+      // Actually, this drain the data out of socket buffer. Although the `buffer_` may
+      // already has the enough data, but we have to drain the socket to ensure there is
+      // no more `Read` event.
+      Buffer::Reservation reservation = buffer_->reserveForRead();
+      auto length_to_read = std::min(reservation.length(), length);
+      // TODO(soulxu): this should be in a while loop until eagain returned.
+      Api::IoCallUint64Result result =
+          readv(length_to_read, reservation.slices(), reservation.numSlices());
+      uint64_t bytes_to_commit = result.ok() ? result.return_value_ : 0;
+      reservation.commit(bytes_to_commit);
+      if (file_event_) {
+        file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
+      }
+      // The read failed, then we return the failure also.
+      if (bytes_to_commit == 0) {
+        return result;
+      }
+      auto copy_size = std::min(buffer_->length(), length);
+      buffer_->copyOut(0, copy_size, buffer);
+      return Api::IoCallUint64Result(copy_size, Api::IoErrorPtr(nullptr, [](Api::IoError*) {}));
+      ;
+    } else {
+      if (buffer_->length() > 0) {
+        auto copy_size = std::min(buffer_->length(), length);
+        buffer_->copyOut(0, copy_size, buffer);
+        buffer_->drain(copy_size);
+        if (copy_size < length) {
+          length = length - copy_size;
+        } else {
+          if (file_event_) {
+            file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
```

Comment:
In general it is assumed that it is a responsibility of the caller to re-active the event if they expect more data unless the error is WOULD_BLOCK

In the case of tls/proxy inspectors they should be able to determine that the data they currently have are not available to do their work and they should re-enable read. This is something that `connection_impl` does well.

Reply:
yea, same as the previous one, whether we should emulate all the cases. or we can just add assert here, assume people always read the data until WOULD_BLOCK return.